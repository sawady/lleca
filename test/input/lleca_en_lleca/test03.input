/* GramÃ¡tica de Lleca escrita en Lleca */

grammar
| rule grammar => Cons($1, $2)
| /*EMPTY*/    => Nil

rule
| ID productions => Rule($1, $2)

productions
| /*EMPTY*/              => Nil
| production productions => Cons($1, $2)

production
| "|" expansion "=>" term => Production($2, $4)

expansion
| /*EMPTY*/        => Nil
| symbol expansion => Cons($1, $2)

symbol
| "ID"     => SymbolId
| "STRING" => SymbolString
| "NUM"    => SymbolNum
| STRING   => SymbolLiteral($1)
| ID       => SymbolNonterm($1)

term
| "_"                  => TermHole
| ID arguments         => TermNode($1, $2)
| STRING               => TermString($1)
| NUM                  => TermNum($1)
| "$" NUM substitution => TermParameter($1, $2)

arguments
| /*EMPTY*/       => Nil
| "(" arglist ")" => $2

arglist
| /*EMPTY*/     => Nil
| term arglistc => Cons($1, $2)

arglistc
| /*EMPTY*/         => Nil
| "," term arglistc => Cons($2, $3)

substitution
| /*EMPTY*/     => Nothing
| "[" term "]"  => Just($2)

