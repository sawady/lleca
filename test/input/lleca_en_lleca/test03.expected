Cons(
  Rule(
    grammar,
    Cons(
      Production(
        Cons(
          SymbolNonterm(
            rule
          ),
          Cons(
            SymbolNonterm(
              grammar
            ),
            Nil
          )
        ),
        TermNode(
          Cons,
          Cons(
            TermParameter(
              $,
              1
            ),
            Cons(
              TermParameter(
                $,
                2
              ),
              Nil
            )
          )
        )
      ),
      Cons(
        Production(
          Nil,
          TermNode(
            Nil,
            Nil
          )
        ),
        Nil
      )
    )
  ),
  Cons(
    Rule(
      rule,
      Cons(
        Production(
          Cons(
            SymbolId,
            Cons(
              SymbolNonterm(
                productions
              ),
              Nil
            )
          ),
          TermNode(
            Rule,
            Cons(
              TermParameter(
                $,
                1
              ),
              Cons(
                TermParameter(
                  $,
                  2
                ),
                Nil
              )
            )
          )
        ),
        Nil
      )
    ),
    Cons(
      Rule(
        productions,
        Cons(
          Production(
            Nil,
            TermNode(
              Nil,
              Nil
            )
          ),
          Cons(
            Production(
              Cons(
                SymbolNonterm(
                  production
                ),
                Cons(
                  SymbolNonterm(
                    productions
                  ),
                  Nil
                )
              ),
              TermNode(
                Cons,
                Cons(
                  TermParameter(
                    $,
                    1
                  ),
                  Cons(
                    TermParameter(
                      $,
                      2
                    ),
                    Nil
                  )
                )
              )
            ),
            Nil
          )
        )
      ),
      Cons(
        Rule(
          production,
          Cons(
            Production(
              Cons(
                SymbolLiteral(
                  "|"
                ),
                Cons(
                  SymbolNonterm(
                    expansion
                  ),
                  Cons(
                    SymbolLiteral(
                      "=>"
                    ),
                    Cons(
                      SymbolNonterm(
                        term
                      ),
                      Nil
                    )
                  )
                )
              ),
              TermNode(
                Production,
                Cons(
                  TermParameter(
                    $,
                    2
                  ),
                  Cons(
                    TermParameter(
                      $,
                      4
                    ),
                    Nil
                  )
                )
              )
            ),
            Nil
          )
        ),
        Cons(
          Rule(
            expansion,
            Cons(
              Production(
                Nil,
                TermNode(
                  Nil,
                  Nil
                )
              ),
              Cons(
                Production(
                  Cons(
                    SymbolNonterm(
                      symbol
                    ),
                    Cons(
                      SymbolNonterm(
                        expansion
                      ),
                      Nil
                    )
                  ),
                  TermNode(
                    Cons,
                    Cons(
                      TermParameter(
                        $,
                        1
                      ),
                      Cons(
                        TermParameter(
                          $,
                          2
                        ),
                        Nil
                      )
                    )
                  )
                ),
                Nil
              )
            )
          ),
          Cons(
            Rule(
              symbol,
              Cons(
                Production(
                  Cons(
                    SymbolLiteral(
                      "ID"
                    ),
                    Nil
                  ),
                  TermNode(
                    SymbolId,
                    Nil
                  )
                ),
                Cons(
                  Production(
                    Cons(
                      SymbolLiteral(
                        "STRING"
                      ),
                      Nil
                    ),
                    TermNode(
                      SymbolString,
                      Nil
                    )
                  ),
                  Cons(
                    Production(
                      Cons(
                        SymbolLiteral(
                          "NUM"
                        ),
                        Nil
                      ),
                      TermNode(
                        SymbolNum,
                        Nil
                      )
                    ),
                    Cons(
                      Production(
                        Cons(
                          SymbolString,
                          Nil
                        ),
                        TermNode(
                          SymbolLiteral,
                          Cons(
                            TermParameter(
                              $,
                              1
                            ),
                            Nil
                          )
                        )
                      ),
                      Cons(
                        Production(
                          Cons(
                            SymbolId,
                            Nil
                          ),
                          TermNode(
                            SymbolNonterm,
                            Cons(
                              TermParameter(
                                $,
                                1
                              ),
                              Nil
                            )
                          )
                        ),
                        Nil
                      )
                    )
                  )
                )
              )
            ),
            Cons(
              Rule(
                term,
                Cons(
                  Production(
                    Cons(
                      SymbolLiteral(
                        "_"
                      ),
                      Nil
                    ),
                    TermNode(
                      TermHole,
                      Nil
                    )
                  ),
                  Cons(
                    Production(
                      Cons(
                        SymbolId,
                        Cons(
                          SymbolNonterm(
                            arguments
                          ),
                          Nil
                        )
                      ),
                      TermNode(
                        TermNode,
                        Cons(
                          TermParameter(
                            $,
                            1
                          ),
                          Cons(
                            TermParameter(
                              $,
                              2
                            ),
                            Nil
                          )
                        )
                      )
                    ),
                    Cons(
                      Production(
                        Cons(
                          SymbolString,
                          Nil
                        ),
                        TermNode(
                          TermString,
                          Cons(
                            TermParameter(
                              $,
                              1
                            ),
                            Nil
                          )
                        )
                      ),
                      Cons(
                        Production(
                          Cons(
                            SymbolNum,
                            Nil
                          ),
                          TermNode(
                            TermNum,
                            Cons(
                              TermParameter(
                                $,
                                1
                              ),
                              Nil
                            )
                          )
                        ),
                        Cons(
                          Production(
                            Cons(
                              SymbolLiteral(
                                "$"
                              ),
                              Cons(
                                SymbolNum,
                                Cons(
                                  SymbolNonterm(
                                    substitution
                                  ),
                                  Nil
                                )
                              )
                            ),
                            TermNode(
                              TermParameter,
                              Cons(
                                TermParameter(
                                  $,
                                  1
                                ),
                                Cons(
                                  TermParameter(
                                    $,
                                    2
                                  ),
                                  Nil
                                )
                              )
                            )
                          ),
                          Nil
                        )
                      )
                    )
                  )
                )
              ),
              Cons(
                Rule(
                  arguments,
                  Cons(
                    Production(
                      Nil,
                      TermNode(
                        Nil,
                        Nil
                      )
                    ),
                    Cons(
                      Production(
                        Cons(
                          SymbolLiteral(
                            "("
                          ),
                          Cons(
                            SymbolNonterm(
                              arglist
                            ),
                            Cons(
                              SymbolLiteral(
                                ")"
                              ),
                              Nil
                            )
                          )
                        ),
                        TermParameter(
                          $,
                          2
                        )
                      ),
                      Nil
                    )
                  )
                ),
                Cons(
                  Rule(
                    arglist,
                    Cons(
                      Production(
                        Nil,
                        TermNode(
                          Nil,
                          Nil
                        )
                      ),
                      Cons(
                        Production(
                          Cons(
                            SymbolNonterm(
                              term
                            ),
                            Cons(
                              SymbolNonterm(
                                arglistc
                              ),
                              Nil
                            )
                          ),
                          TermNode(
                            Cons,
                            Cons(
                              TermParameter(
                                $,
                                1
                              ),
                              Cons(
                                TermParameter(
                                  $,
                                  2
                                ),
                                Nil
                              )
                            )
                          )
                        ),
                        Nil
                      )
                    )
                  ),
                  Cons(
                    Rule(
                      arglistc,
                      Cons(
                        Production(
                          Nil,
                          TermNode(
                            Nil,
                            Nil
                          )
                        ),
                        Cons(
                          Production(
                            Cons(
                              SymbolLiteral(
                                ","
                              ),
                              Cons(
                                SymbolNonterm(
                                  term
                                ),
                                Cons(
                                  SymbolNonterm(
                                    arglistc
                                  ),
                                  Nil
                                )
                              )
                            ),
                            TermNode(
                              Cons,
                              Cons(
                                TermParameter(
                                  $,
                                  2
                                ),
                                Cons(
                                  TermParameter(
                                    $,
                                    3
                                  ),
                                  Nil
                                )
                              )
                            )
                          ),
                          Nil
                        )
                      )
                    ),
                    Cons(
                      Rule(
                        substitution,
                        Cons(
                          Production(
                            Nil,
                            TermNode(
                              Nothing,
                              Nil
                            )
                          ),
                          Cons(
                            Production(
                              Cons(
                                SymbolLiteral(
                                  "["
                                ),
                                Cons(
                                  SymbolNonterm(
                                    term
                                  ),
                                  Cons(
                                    SymbolLiteral(
                                      "]"
                                    ),
                                    Nil
                                  )
                                )
                              ),
                              TermNode(
                                Just,
                                Cons(
                                  TermParameter(
                                    $,
                                    2
                                  ),
                                  Nil
                                )
                              )
                            ),
                            Nil
                          )
                        )
                      ),
                      Nil
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )
)
